abci/proposals/validate.go

```diff
@@ -28,33 +32,80 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
 
 	// Validate all oracle vote extensions.
 	for _, vote := range extendedCommitInfo.Votes {
-		address := sdk.ConsAddress{}
-		if err := address.Unmarshal(vote.Validator.Address); err != nil {
+		address := sdk.ConsAddress(vote.Validator.Address)
+		// The vote extension are from the previous block.
+		if err := validateVoteExtension(ctx, vote, h.voteExtensionCodec, h.currencyPairStrategy); err != nil {
 			h.logger.Error(
-				"failed to unmarshal validator address",
+				"failed to validate oracle vote extension",
 				"height", height,
+				"validator", address.String(),
+				"err", err,
 			)
 
 			return err
 		}
+	}
 
-		voteExt, err := h.voteExtensionCodec.Decode(vote.VoteExtension)
-		if err != nil {
-			return err
-		}
+	return nil
+}
 
-		// The vote extension are from the previous block.
-		if err := ve.ValidateOracleVoteExtension(ctx, voteExt, h.currencyPairStrategy); err != nil {
-			h.logger.Error(
-				"failed to validate oracle vote extension",
-				"height", height,
-				"validator", address.String(),
+// PruneAndValidateExtendedCommitInfo validates each vote-extension in the extended commit, and removes
+// any vote-extensions that are invalid. Removal will effectively treat the validator's
+// vote as absent.  This function performs all validation that ValidateExtendedCommitInfo performs.
+func (h *ProposalHandler) PruneAndValidateExtendedCommitInfo(
+	ctx sdk.Context, extendedCommitInfo cometabci.ExtendedCommitInfo,
+) (cometabci.ExtendedCommitInfo, error) {
+	// Validate all oracle vote extensions.
+	for i, vote := range extendedCommitInfo.Votes {
+		// validate the vote-extension
+		if err := validateVoteExtension(ctx, vote, h.voteExtensionCodec, h.currencyPairStrategy); err != nil {
+			h.logger.Info(
+				"failed to validate vote extension - pruning vote",
 				"err", err,
+				"validator", vote.Validator.Address,
 			)
 
-			return err
+			// failed to validate this vote-extension, mark it as absent in the original commit
+			vote.BlockIdFlag = cometproto.BlockIDFlagAbsent
+			vote.ExtensionSignature = nil
+			vote.VoteExtension = nil
+			extendedCommitInfo.Votes[i] = vote
 		}
 	}
 
+	// validate after pruning
+	if err := h.validateVoteExtensionsFn(ctx, extendedCommitInfo); err != nil {
```

Comment:
It's not clear to the user whether or not this method always checks that the votes are a quorum, no? Why not perform this check in the pruning fn?

Reply:
what do you mean? This is in the pruning function.

We:

1. Prune invalid VEs and set them to absent + nil
2. check that _after_ pruning, is there still quorum

Reply:
Ya, I'm asking why we're not explicitly checking that there's a quorum here. Instead it looks like we expect that the `ValidateVoteExtensionsFn` does this for us?

Reply:
Actually, this is fine, we use `validateVoteExtensionsFn` in process anyway, so this check makes sense

Reply:
small nit, it is a bit confusing read this code and seeing `validateVoteExtension` and `validateVoteExtensionsFn`